### PR TITLE
improve(price-feed): Add BasisPriceFeed which computes prices for BASIS-esque identifiers

### PR DESCRIPTION
**Pull request by chicfilabae**
Per the requirements of the [new price identifier umip template](https://github.com/UMAprotocol/UMIPs/pull/123/files), this is a companion PR to https://github.com/UMAprotocol/UMIPs/pull/140.

This implementation heavily derives from @nicholaspai [PR](https://github.com/UMAprotocol/protocol/pull/2354/). @nicholaspai since there is a large amount of redundancy of code between this and `BasketSpreadPriceFeed.js`, feel free to abstract away as you care to. Aside from the lower/upper bounds as params, can either specify an fxn as a param for how to combine both baseline and experimental pricefeeds, or just have a flag to decide whether to normalize or not, etc. 

At any rate, per the template this is meant to be "...with an example price feed that inherits from PriceFeedInterface", and as such, here it is.

### CODE REVIEW

File: packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js
Comment (daywiss) [inline packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js:77]: just for reference, the umip formula: `min(max(100.0 * [1 + (F-S)/S], 75.0), 125.0)`
I dont see here where you are adding `1` to the spread value, maybe its not needed though im not sure. what do you think?

**Review by daywiss — DISMISSED** [review]
looks good. will approve assuming you will address comments

File: packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js
Comment (daywiss) [inline packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js:83]: ci is failing because its missing an `)` same with the else if

File: packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js
Comment (chicfilabae) [reply packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js:77]: great catch. the PR was updated after this implementation so was out of sync. making changes now.

Comment (chicfilabae) [reply packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js]: Done

File: packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js
Comment (chicfilabae) [reply packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js]: Done

File: packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js
Comment (chicfilabae) [reply packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js:63]: This exists in `BasketSpreadPriceFeed.js` from @nicholaspai, didn't modify that logic

**Comment by chicfilabae** [comment]
@daywiss Just updated!

**Comment by daywiss** [comment]
@UMAprotocol/eng do we need to include tests in this PR?

**Comment by daywiss** [comment]
@chicfilabae it looks like there are some linter errors in ci, we have a husky pre commit hook that usually runs and fixes this automatically, but sometimes it  doesnt run.  you can try running yarn `lint-fix` in the protocol root and see if that works.

**Comment by chrismaree** [comment]
Thanks for the PR! I have a few questions/points:

**1**) Your UMIP refers to a "basket of BTC futures" and pulls this information from FTX, Binance & Okex. For this to work you will need to define a set of inputs to your `BasisPriceFeed`. Fortunately, all futures you refer to are supported by cryptowatch, which means you can re-use the feeds from there. I went and found the feeds for you, they are here:

FTX
1) https://cryptowat.ch/charts/FTX:BTC-USD
2) https://cryptowat.ch/charts/FTX:BTC-USD-FUTURE-QUARTERLY-26MAR21
3) https://cryptowat.ch/charts/FTX:BTC-USD-FUTURE-QUARTERLY-25JUN21

Binance: 
1) https://cryptowat.ch/charts/BINANCE-US:BTC-USD
2) https://cryptowat.ch/charts/BINANCE:BTC-USD-FUTURE-INVERSE-QUARTERLY-26MAR21
3) https://cryptowat.ch/charts/BINANCE:BTC-USD-FUTURE-INVERSE-QUARTERLY-25JUN21

Note this is an inverse feed. I could not find a non-inverse feed that had the right expiration. maybe you can?

OKex:
1) https://cryptowat.ch/charts/OKEX:BTC-USDC
2) https://cryptowat.ch/charts/OKEX:BTC-USD-FUTURE-INVERSE-QUARTERLY
3) https://cryptowat.ch/charts/OKEX:BTC-USD-FUTURE-INVERSE-BIQUARTERLY

Note i) the spot market here is USDC. I think this is fine as they don't seem to have a USD spot market ii) this is also inverse feeds. iii) they don't have exact dated expiration for futures and only seem to have "QUARTERLY" and "BIQUARTERLY".

Using these as inputs, you need to define a default config to parameterized your feed. This would act similarly to how `STABLESPREAD/USDC` would work, for example. See https://github.com/UMAprotocol/protocol/blob/95fb302f5b370658adced9cf23c3f897fe00d7d5/packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js#L144 for what I'm referring to. You can pull out the `exchange` and `pair` part from the URLs I linked above.

**3** you need to add changes to `CreatePriceFeed.js` to enable the bots to construct your new feed. Seeing this feed is very similar in form to the `basketSpread` feed, you should be able to mimic what's happening on that feed, as can be seen [here](https://github.com/UMAprotocol/protocol/blob/95fb302f5b370658adced9cf23c3f897fe00d7d5/packages/financial-templates-lib/src/price-feed/CreatePriceFeed.js#L142).

**3** your code is not yet linted correctly to be merged into the repo. 

Once you've addressed the above 3 issues we can look at getting this merged. Thanks again for your contribution, excited for this product to launch!

**Comment by chicfilabae** [comment]
@daywiss @chrismaree 

made all the above changes w.r.t. tests, linting, default price feeds, etc.

File: package.json
Comment (mrice32) [inline package.json:32]: Could you run `yarn` from the root of the repo and commit any changes in `yarn.lock`?

File: packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js
Comment (mrice32) [inline packages/financial-templates-lib/src/price-feed/BasisPriceFeed.js:97]: We just made a change on master changing this function to async, so you'll need to make some minor changes here.

```suggestion
  async getHistoricalPrice(time) {
    const errors = [];
    const futurePrices = await Promise.all(this.futurePriceFeeds.map(priceFeed => priceFeed.getHistoricalPrice(time).catch(err => errors.push(err))));
    const spotPrices = await Promise.all(this.spotPriceFeeds.map(priceFeed => priceFeed.getHistoricalPrice(time).catch(err => errors.push(err))));
    
    if (errors.length > 0) {
      throw errors;
    }
    
    return this._getSpreadFromBasketPrices(futurePrices, spotPrices);
  }
```

File: packages/financial-templates-lib/test/price-feed/BasisPriceFeed.js
Comment (mrice32) [inline packages/financial-templates-lib/test/price-feed/BasisPriceFeed.js:168]:
```suggestion
      assert.equal(await basisPriceFeed.getHistoricalPrice(arbitraryHistoricalTimestamp), toWei("200"));
```

File: packages/financial-templates-lib/test/price-feed/BasisPriceFeed.js
Comment (mrice32) [inline packages/financial-templates-lib/test/price-feed/BasisPriceFeed.js:186]:
```suggestion
        await basisPriceFeed.getHistoricalPrice(arbitraryHistoricalTimestamp),
```

File: packages/financial-templates-lib/test/price-feed/BasisPriceFeed.js
Comment (mrice32) [inline packages/financial-templates-lib/test/price-feed/BasisPriceFeed.js:319]:
```suggestion
      assert.equal(await basisPriceFeed.getHistoricalPrice(arbitraryHistoricalTimestamp), toWei("130"));
```

File: packages/financial-templates-lib/test/price-feed/BasisPriceFeed.js
Comment (mrice32) [inline packages/financial-templates-lib/test/price-feed/BasisPriceFeed.js:338]:
```suggestion
        await basisPriceFeed.getHistoricalPrice(arbitraryHistoricalTimestamp),
```

File: packages/financial-templates-lib/test/price-feed/BasisPriceFeed.js
Comment (mrice32) [inline packages/financial-templates-lib/test/price-feed/BasisPriceFeed.js:244]:
```suggestion
      assert.equal(await basisPriceFeed.getHistoricalPrice(arbitraryHistoricalTimestamp), toWei("70"));
```

File: packages/financial-templates-lib/test/price-feed/BasisPriceFeed.js
Comment (mrice32) [inline packages/financial-templates-lib/test/price-feed/BasisPriceFeed.js:262]:
```suggestion
        await basisPriceFeed.getHistoricalPrice(arbitraryHistoricalTimestamp),
```

**Review by mrice32 — COMMENTED** [review]
We made a minor change to the price feed interface that will require a few changes on your end. I've made (untested) suggestions that should make this PR compatible. Sorry about the last minute whiplash!

**Comment by chicfilabae** [comment]
@mrice32 Changes you requested above have been made

**Review by mrice32 — APPROVED** [review]
Looks great!

One minor issue you will need to resolve before merging.

There are conflicts in the yarn.lock file. You don't want to merge this manually, as that's usually a disaster for an autogenerated file like that. What I would suggest doing is:
1. Make sure you have the latest master pulled down locally.
2. Run `git checkout master yarn.lock` from your branch.
3. Run `git merge master` from your branch to merge in recent changes to master.
4. Run `yarn` from your branch to reconstruct the yarn.lock file _on top_ of the lockfile changes that you pulled in from master.

Should be pretty quick and easy. Once those conflicts are gone, I think we should be ready to merge! Will ask @chrismaree and @daywiss to take a final look, though!

**Review by chrismaree — COMMENTED** [review]
Looks like you've got some failing tests in CI: https://app.circleci.com/pipelines/github/UMAprotocol/protocol/10678/workflows/a61a1706-a313-45fe-8727-4a7741cbee03/jobs/75549

**Comment by chrismaree** [comment]
> @chrismaree was Point 1 of your comment here: [#2457 (comment)](https://github.com/UMAprotocol/protocol/pull/2457#issuecomment-769047831) addressed satisfactorily?

It looks correct, assuming the underlying futures feeds are right. I'm not 100% sure why some of the feeds are indicated as inverse feeds on cryptowatch. For example `btcusd-quarterly-future-inverse` is an inverse feed on `okex` but `btcusd-quarterly-futures-26mar21` is not an inverse feed on `ftx`. @chicfilabae is there any extra computation needed to be done on these feeds?

Also, I'd love to run this locally to double-check the output and execution of the feeds. can do that once we've got it merged.

File: packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js
Comment (daywiss) [inline packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js:299]: is there a reason why these all use a medianizer feed rather than just the cryptowatch feed directly?

**Comment by chicfilabae** [comment]
Closing this in favor of specifying this in config land with the new expression price feed functionality Matt has implemented.

**Pull request closed without merging**
